### PR TITLE
Flink Delta Sink - PR 3

### PR DESCRIPTION
**Pull request by pkubit-g**

### CODE REVIEW

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java [hunk 1]
```diff
@@ -105,11 +113,54 @@
...
         writer.initializeState(states);
         return writer;
     }
 
+    /**
+     * In order to gurantee the idempotency of the GlobalCommitter we need unique identifier of the
+     * app. We obtain it with simple logic: if it's the first run of the application (so no restart
+     * from snapshot or failure recovery happened and the writer's state is empty) then assign appId
+     * to a newly generated UUID that will be further stored in the state of each writer.
+     * Alternatively if the writer's states are not empty then we resolve appId from on of the
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java]: nit: `from one of`

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSinkBuilder.java [hunk 2]
```diff
@@ -71,6 +76,13 @@
      */
     private final Path tableBasePath;
 
+    /**
+     * Unique identifier of the current Flink's app. Value from this builder will be read
+     * only during the fresh start of the application. For restarts or failure recovery
+     * it will be resolved from the snaphosted state.
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSinkBuilder.java]: nit: `snapshoted`

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java [hunk 3]
```diff
@@ -96,10 +111,15 @@
...
-     * @param basePath       the base path for the table
-     * @param bucketAssigner The {@link BucketAssigner} provided by the user.
-     * @param bucketWriter   The {@link DeltaBulkBucketWriter} to be used when writing data.
-     * @param rollingPolicy  The {@link CheckpointRollingPolicy} as specified by the user.
+     * @param basePath         The base path for the table
+     * @param bucketAssigner   The {@link BucketAssigner} provided by the user.
+     * @param bucketWriter     The {@link DeltaBulkBucketWriter} to be used when writing data.
+     * @param rollingPolicy    The {@link CheckpointRollingPolicy} as specified by the user.
+     * @param appId            Unique identifier of the current Flink app. This identifier needs to
+     *                         constant across all app's restarts to guarantee idempotent
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java]: nit: `needs to be constant`

**Comment by scottsand-db** [comment]
@pkubit-g can you add a brief PR description?

**Review by scottsand-db — APPROVED** [review]
LGTM - good catch + fix using the next checkpoint ID.
left grammar nits

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSinkBuilder.java [hunk 4]
```diff
@@ -113,14 +126,28 @@ protected DeltaSinkBuilder(
...
         this.serializableConfiguration = new SerializableConfiguration(checkNotNull(conf));
         this.bucketCheckInterval = bucketCheckInterval;
         this.writerFactory = writerFactory;
         this.bucketAssigner = checkNotNull(assigner);
         this.rollingPolicy = checkNotNull(policy);
         this.outputFileConfig = checkNotNull(outputFileConfig);
+        this.appId = appId;
+    }
+
+    public DeltaSinkBuilder<IN> withAppId(final String appId) {
```
Comment (tdas) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSinkBuilder.java]: where is this used?

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/committables/DeltaCommittable.java [hunk 5]
```diff
@@ -44,11 +44,27 @@
 
     private final DeltaPendingFile deltaPendingFile;
 
-    public DeltaCommittable(DeltaPendingFile deltaPendingFile) {
+    private final String appId;
+
+    private final long checkpointId;
```
Comment (tdas) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/committables/DeltaCommittable.java:81]: docs to define exactly what they are.

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java [hunk 6]
```diff
@@ -63,6 +64,20 @@
...
+     */
+    public static final String NOOP_WRITER_STATE = "<noop-writer-state>";
+
+    ///////////////////////////////////////////////////////////////////////////
+    // DeltaSink-specific fields
+    ///////////////////////////////////////////////////////////////////////////
+
+    private final String appId;
+
+    private long nextCheckpointId;
```
Comment (tdas) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java:105]: add docs explaining what these 2 are

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java [hunk 7]
```diff
@@ -63,6 +64,20 @@
```
Comment (tdas) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java:76]: ~Can we rename this class tot DeltaSinkWriter? i think that makes it more consistent with the pattern other classes follow: DeltaX extends X~
nvm, i can see that all the other classes like DeltaWriterBucketState ... all of them have to renamed... and they are going to get longer. so ignore this.

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucketState.java [hunk 8]
```diff
@@ -47,10 +47,18 @@
```
Comment (tdas) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucketState.java:57]: update docs saying that `extends the state by adding application's unique identifier and checkpoint interval identifier to guarantee....`

File: flink-connector/src/test/java/org/apache/flink/connector/delta/sink/committables/DeltaCommittableSerializerTest.java [hunk 9]
```diff
@@ -43,16 +43,9 @@ public void testCommittableWithPendingFileForNonPartitionedTable() throws IOExce
...
-            TestDeltaCommittable.getTestDeltaCommittableWithPendingFile();
-
-        // WHEN
-        DeltaCommittable deserialized = serializeAndDeserialize(committable);
-
-        // THEN
-        TestDeltaCommittable.validateDeltaCommittablesEquality(committable, deserialized);
+    public void testCommittableWithPendingFileForPartitionedTable() {
+        // TODO
+        // will be added in next PRs
```
Comment (tdas) [inline flink-connector/src/test/java/org/apache/flink/connector/delta/sink/committables/DeltaCommittableSerializerTest.java]: ???

**Pull request closed without merging**